sql/hive/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala

```diff
@@ -0,0 +1,110 @@
+
+        val attrs = df.select("j", "k").schema.toAttributes
+        val checkBucketId = rdd.mapPartitionsWithIndex((index, rows) => {
+          val bucketIdExpression = BucketingUtils.bucketIdExpression(attrs, 8)
+          val projection = UnsafeProjection.create(bucketIdExpression :: Nil, attrs)
+          rows.map(row => projection(row).getInt(0) == index)
+        })
+
+        assert(checkBucketId.collect().reduce(_ && _))
+      }
+    }
+  }
+
+  test("avoid shuffle and sort when join 2 bucketed tables") {
+    def createTable(df: DataFrame, tableName: String): Unit = {
+      df.write.format("parquet").bucketBy(8, "i", "j").sortBy("i", "j").saveAsTable(tableName)
+    }
+
+    withTable("bucketed_table1", "bucketed_table2") {
+      val df1 = (0 until 50).map(i => (i % 5, i % 13, i.toString)).toDF("i", "j", "k")
+      val df2 = (0 until 50).map(i => (i % 7, i % 11, i.toString)).toDF("i", "j", "k")
+
+      createTable(df1, "bucketed_table1")
+      createTable(df2, "bucketed_table2")
+
+      withSQLConf("spark.sql.autoBroadcastJoinThreshold" -> "0") {
+        val t1 = hiveContext.table("bucketed_table1")
+        val t2 = hiveContext.table("bucketed_table2")
+
+        val joined = t1.join(t2, t1("i") === t2("i") && t1("j") === t2("j"))
+        assert(joined.queryExecution.executedPlan.isInstanceOf[SortMergeJoin])
+        assert(joined.queryExecution.executedPlan.find(_.isInstanceOf[Exchange]).isEmpty)
+        assert(joined.queryExecution.executedPlan.find(_.isInstanceOf[Sort]).isEmpty)
+        checkAnswer(
+          joined.sort(t1("k")),
+          df1.join(df2, df1("i") === df2("i") && df1("j") === df2("j")).sort(df1("k")))
+      }
+    }
+  }
+
+  test("shuffle when join 2 bucketed tables with bucketing disabled") {
+    withSQLConf("spark.sql.sources.bucketing.enabled" -> "false") {
+      def createTable(df: DataFrame, tableName: String): Unit = {
+        df.write.format("parquet").bucketBy(8, "i", "j").saveAsTable(tableName)
+      }
+
+      withTable("bucketed_table1", "bucketed_table2") {
+        val df1 = (0 until 50).map(i => (i % 5, i % 13, i.toString)).toDF("i", "j", "k")
+        val df2 = (0 until 50).map(i => (i % 7, i % 11, i.toString)).toDF("i", "j", "k")
+
+        createTable(df1, "bucketed_table1")
+        createTable(df2, "bucketed_table2")
+
+        withSQLConf("spark.sql.autoBroadcastJoinThreshold" -> "0") {
+          val t1 = hiveContext.table("bucketed_table1")
+          val t2 = hiveContext.table("bucketed_table2")
+
+          val joined = t1.join(t2, t1("i") === t2("i") && t1("j") === t2("j"))
+          assert(joined.queryExecution.executedPlan.isInstanceOf[SortMergeJoin])
+          assert(joined.queryExecution.executedPlan.find(_.isInstanceOf[Exchange]).isDefined)
+          checkAnswer(
+            joined.sort(t1("k")),
+            df1.join(df2, df1("i") === df2("i") && df1("j") === df2("j")).sort(df1("k")))
+        }
+      }
+    }
+  }
+}
```

Comment:
Add a case of 1 bucket and 1 not
Add a case of bucketed but on different things